File: sqldelight-sample/src/main/java/com/example/sqldelight/hockey/data/Team.java
package com.example.sqldelight.hockey.data;

import com.google.auto.value.AutoValue;
import com.squareup.sqldelight.RowMapper;
import java.util.Calendar;

@AutoValue
public abstract class Team implements TeamModel {
  private static final DateAdapter DATE_ADAPTER = new DateAdapter();

  public static final Factory<Team> FACTORY = new Factory<>(new Creator<Team>() {
    @Override public Team create(long Id, String name, Calendar founded, String coach, Long captain,
        boolean wonCup) {
      return new AutoValue_Team(Id, name, founded, coach, captain, wonCup);
    }
  }, DATE_ADAPTER);

  public static final RowMapper<Team> MAPPER = FACTORY.select_allMapper();

  public static Marshal marshal() {
    return new Marshal(DATE_ADAPTER);
  }
}

File: sqldelight-core-integration-tests/src/test/kotlin/com/example/Team.kt
package com.example

import com.squareup.sqldelight.ColumnAdapter
import com.squareup.sqldelight.core.integration.Shoots
import kotlin.Long
import kotlin.String

interface Team {
    val name: String

    val captain: Long

    val inner_type: Shoots.Type?

    val coach: String

    class Adapter(internal val inner_typeAdapter: ColumnAdapter<Shoots.Type, String>)

    data class Impl(
            override val name: String,
            override val captain: Long,
            override val inner_type: Shoots.Type?,
            override val coach: String
    ) : Team
}

abstract class TeamModel : Team {
    final override val name: String
        get() = name()

    final override val captain: Long
        get() = captain()

    final override val inner_type: Shoots.Type?
        get() = inner_type()

    final override val coach: String
        get() = coach()

    abstract fun name(): String

    abstract fun captain(): Long

    abstract fun inner_type(): Shoots.Type?

    abstract fun coach(): String
}
